mountpoint-s3/tests/fuse_tests/consistency_test.rs

```diff
@@ -83,3 +83,98 @@ fn page_cache_sharing_test_mock_with_cache(prefix: &str) {
         prefix,
     );
 }
+
+#[cfg(target_os = "linux")]
+mod direct_io {
+    use super::*;
+
+    use std::fs::OpenOptions;
+    use std::os::unix::fs::OpenOptionsExt;
+    use std::time::Duration;
+
+    use test_case::test_case;
+
+    use mountpoint_s3::fs::{CacheConfig, S3FilesystemConfig};
+
+    fn cache_and_direct_io_test<F>(creator_fn: F, prefix: &str)
+    where
+        F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
+    {
+        const OBJECT_SIZE: usize = 8;
+
+        let test_session_conf = TestSessionConfig {
+            filesystem_config: S3FilesystemConfig {
+                cache_config: CacheConfig {
+                    serve_lookup_from_cache: true,
+                    dir_ttl: Duration::from_secs(600),
+                    file_ttl: Duration::from_secs(600),
+                },
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+        let (mount_point, _session, mut test_client) = creator_fn(prefix, test_session_conf);
+
+        let file_name = "file.bin";
+
+        // Create the first version of the file
+        let old_contents = vec![0x0fu8; OBJECT_SIZE];
+        test_client.put_object(file_name, &old_contents).unwrap();
+
+        // Open and read fully the file before updating it remotely
+        let old_file = File::open(mount_point.path().join(file_name)).unwrap();
+        let mut buf = vec![0u8; OBJECT_SIZE];
+        old_file.read_exact_at(&mut buf, 0).unwrap();
+        assert_eq!(buf, &old_contents[..buf.len()]);
+
+        let new_contents = vec![0xffu8; OBJECT_SIZE];
+        test_client.put_object(file_name, &new_contents).unwrap();
+
+        // Open the file again, which should be reading from cache
+        for _ in 0..2 {
+            let new_file = File::open(mount_point.path().join(file_name)).unwrap();
+            new_file
+                .read_exact_at(&mut buf, 0)
+                .expect("should be OK as result is cached");
+            assert_eq!(
+                buf,
+                &old_contents[..buf.len()],
+                "bytes read should be old object from cache"
+            );
+        }
+
+        // Open the file w/ O_DIRECT, which should see the new file on S3 despite the old file being cached
+        let mut buf = [0u8; OBJECT_SIZE];
```

Comment:
If I use `vec![0u8; OBJECT_SIZE]` here, I sometimes find that the buffer hasn't been written to (even though `read_exact_at` has returned `Ok`). This makes me super uncomfortable (which is why this PR is still draft, not ready to merge).

Reply:
`O_DIRECT` is really hard to test under the Rust test runner because it has undefined behavior if run concurrently with a `fork`. We've seen it before in #114. Do you still see the issue if you filter to run only these tests?

Reply:
Ah, I saw those limitations with `O_DIRECT` and `fork` but wasn't expecting us to be using it in the tests. That sounds like a probable cause!

I don't see the issue when running the test on its own.